bftengine/include/bftengine/ReservedPages.hpp

```diff
@@ -0,0 +1,116 @@
+// Concord
+//
+// Copyright (c) 2018 VMware, Inc. All Rights Reserved.
+//
+// This product is licensed to you under the Apache 2.0 license (the "License").
+// You may not use this product except in compliance with the Apache 2.0
+// License.
+//
+// This product may include a number of subcomponents with separate copyright
+// notices and license terms. Your use of these subcomponents is subject to the
+// terms and conditions of the subcomponent's license, as noted in the LICENSE
+// file.
+
+#pragma once
+
+#include <cstdint>
+#include <string>
+#include <typeindex>
+#include <algorithm>
+#include <tuple>
+#include <vector>
+#include <iostream>
+#include "demangle.hpp"
+
+namespace bftEngine {
+
+class ReservedPages {
+ public:
+  static uint32_t totalNumberOfPages() {
+    uint32_t numPages{0};
+    for (auto& elt : registry()) numPages += std::get<2>(elt);
+
+    return numPages;
+  }
+
+ protected:
+  // [idx, typeid, numberOfPages]
+  typedef std::vector<std::tuple<std::uint8_t, std::type_index, std::uint32_t>> Registry;
+
+  static Registry& registry() {
+    static Registry registry_;
+    return registry_;
+  }
+};
+/**
+ * Facility for registering Reserved Pages clients
+ *
+ * Become a Reserved Pages client by sub-classing:
+ * class MyClass: public ResPagesClient<MyClass, idx, requiredNumberOfPages> {};
+ * where idx is a unique integer used for dividing a reserved pages space into sub-spaces.
+ * If requiredNumberOfPages is not known at compile time call setNumResPages(requiredNumberOfPages).
+ */
+template <typename T, uint8_t Idx, uint32_t NumPages = 0>
+class ResPagesClient : public ReservedPages {
+ public:
+  ResPagesClient() { (void)registered_; }
+
+  static bool registerT() {
+    Registry& reg = registry();
+    auto it = std::find_if(reg.begin(), reg.end(), [](const Registry::value_type& v) { return std::get<0>(v) == Idx; });
+    if (it != reg.end()) {
+      std::cerr << "BUG: ResPagesClient<" << demangler::demangle(typeid(T).name()) << ", " << (int)Idx
+                << ">: index is used by " << demangler::demangle(std::get<1>(*it).name());
+      std::terminate();
+    }
+    reg.push_back(Registry::value_type(Idx, std::type_index(typeid(T)), NumPages));
+    // sorting by clients' index
+    std::sort(reg.begin(), reg.end(), [](const Registry::value_type& v1, const Registry::value_type& v2) {
```

Comment:
What is the purpose of this sorting?

Reply:
Sorting by index to enforce offsets order.
Since the registration takes place before main, its order is somewhat arbitrary. I want to enforce the order given by index, i.e. if `ClientsManager : public ResPageClient<ClientsManager, 0>` then its offset is 0.
If `KeyManager: public ResPageClient<KeyManager,25>` then its offset is the number of pages used by ResPagesClients with indices smaller than its.